### PR TITLE
Extend SmoothQuant support (exclude nodes, fuse into layernorm)

### PR DESCRIPTION
**Pull request by fxmarty**
## Type of Change

As per title.

## Description

This PR includes two new features for SmoothQuant (that I was too lazy to split into two PRs):
* Add the fusion of `MatMul -> Add -> MatMul` into `MatMul -> Add` (that is typically the case for layernorm) as https://github.com/mit-han-lab/smoothquant/blob/78badc0d975506de9fe44b2fe79d9a35d0fd4914/smoothquant/smooth.py#L46
* Add a parameter `nodes_to_exclude` following ORT quantizer & Optimum quantizer fashion, to allow to exclude some nodes from being smoothed. This is following https://github.com/mit-han-lab/smoothquant/issues/15#issuecomment-1353390283 (out_proj & fc2 should not be smoothed out to reproduce the paper results).

## How has this PR been tested?

Locally - I did not test that output from the fusion match but I just reused the code from the `mul` method. Let me know if I should add tests.

### CODE REVIEW

**Comment by fxmarty** [comment]
@chensuyue @mengniwang95 @PenghuiCheng @xin3he happy to get a review on this one!

**Comment by chensuyue** [comment]
There are some issues detected by CI, could you have a look?   
https://dev.azure.com/lpot-inc/neural-compressor/_build/results?buildId=21391&view=logs&j=e5896b99-a49d-517b-218b-3b918f0c116d&t=b18c099a-26ee-5571-9980-67a803d9b7da&l=20474   
https://dev.azure.com/lpot-inc/neural-compressor/_build/results?buildId=21391&view=logs&j=c6aa4c58-99e4-54e9-e3eb-cd322b75c938&t=dd753db2-b82a-57eb-728f-6b88742237f1&l=6311

**Comment by fxmarty** [comment]
Thank you @chensuyue, will have a look!

File: neural_compressor/adaptor/ox_utils/smooth_quant.py [hunk 1]
```diff
@@ -145,6 +146,7 @@ def transform(
         calib_iter=100,
         quantize_config=None,
         auto_alpha_args={"alpha_min": 0.3, "alpha_max": 0.7, "alpha_step": 0.05, "attn_method": "min"},
+        nodes_to_exclude: Optional[List[str]] = None,
```
Comment (mengniwang95) [inline neural_compressor/adaptor/ox_utils/smooth_quant.py:149]: how is nodes_to_exclude setting passed from user facing API of neural-compressor?

File: neural_compressor/adaptor/ox_utils/smooth_quant.py [hunk 2]
```diff
@@ -330,6 +336,37 @@ def conv(node, scale):  # pragma: no cover
                 )
             return True
 
+        def mul_add(node, scale):  # pragma: no cover
+            node_parent = self.model.get_parent(node, 0)
+            if not len(self.model.get_parents(node)) == 1 or node_parent.op_type != "Mul":
```
Comment (mengniwang95) [inline neural_compressor/adaptor/ox_utils/smooth_quant.py:341]: below comment said "Add has itself a MatMul before" but here is "node_parent.op_type != 'Mul'", could you align it?

**Comment by mengniwang95** [comment]
Further, could you add a ut to make sure it can work correctly?

**Comment by chensuyue** [comment]
@fxmarty would you like to follow this PR?

**Comment by chensuyue** [comment]
We will have code freeze on 11/22, if this PR could be merged before the date, it can be packaged into v2.4 release.

**Comment by chensuyue** [comment]
@fxmarty will you fix the PR?

**Comment by fxmarty** [comment]
@chensuyue Sorry I did not get time to fix it, I won't be able before the release unfortunately.

**Pull request closed without merging**